### ISSUE TITLE
environment variables are missing

### ISSUE THREAD

**Issue by chack05** [jupyterhub/jupyterhub#1695]
Hi,
I have installed Jupyterhub and a notebook with the default IPython kernel for Python 3 and another one for Python 2.

Now I have this problem that environment variables set in `/etc/bash.bashrc` ~~work in the Python 3 kernel~~, but not in the Python 2 kernel, for all users. Outside of Jupyter(hub) the variables are available in Python 2 as well.

**EDIT**: The variables are missing for Python 3 as well. My mistake.

Jupyterhub version is 0.8.1, OS is Ubuntu 16 LTS.

`jupyter troubleshoot` outputs the following:

```
$PATH:
        /usr/local/bin
        /home/c_hack05/bin
        /home/c_hack05/.local/bin
        /opt/pcraster/bin
        /usr/local/sbin
        /usr/local/bin
        /usr/sbin
        /usr/bin
        /sbin
        /bin
        /usr/games
        /usr/local/games
        /snap/bin

sys.path:
        /usr/local/bin
        /opt/pcraster/python
        /home/c_hack05
        /usr/lib/python2.7
        /usr/lib/python2.7/plat-x86_64-linux-gnu
        /usr/lib/python2.7/lib-tk
        /usr/lib/python2.7/lib-old
        /usr/lib/python2.7/lib-dynload
        /usr/local/lib/python2.7/dist-packages
        /usr/lib/python2.7/dist-packages

sys.executable:
        /usr/bin/python2

sys.version:
        2.7.12 (default, Dec  4 2017, 14:50:18) 
        [GCC 5.4.0 20160609]

platform.platform():
        Linux-4.4.0-116-generic-x86_64-with-Ubuntu-16.04-xenial

which -a jupyter:
        /usr/local/bin/jupyter
        /usr/local/bin/jupyter

pip list:
        backports-abc (0.5)
        backports.shutil-get-terminal-size (1.0.0)
        certifi (2018.1.18)
        decorator (4.2.1)
        enum34 (1.1.6)
        gyp (0.1)
        ipykernel (4.8.2)
        ipython (5.5.0)
        ipython-genutils (0.2.0)
        jupyter-client (5.2.2)
        jupyter-core (4.4.0)
        numpy (1.11.0)
        pathlib2 (2.3.0)
        pexpect (4.4.0)
        pickleshare (0.7.4)
        pip (9.0.1)
        prompt-toolkit (1.0.15)
        ptyprocess (0.5.2)
        Pygments (2.2.0)
        python-dateutil (2.6.1)
        pyzmq (17.0.0)
        scandir (1.7)
        setuptools (20.7.0)
        simplegeneric (0.8.1)
        singledispatch (3.4.0.3)
        six (1.11.0)
        tornado (4.5.3)
        traitlets (4.3.2)
        wcwidth (0.1.7)
        wheel (0.29.0)
```

It even includes the `/opt/pcraster/*` paths.

The env_keep setting in the Jupyterhub configuration is the default value. Jupyterhub, Jupyter etc. are installed at /usr/local and otherwise work for all users.

Is this a bug, my error in the installation process or is there a configuration for the Python 2 kernel I need to customize?

**Comment by minrk** [comment]
By default, a full shell is not invoked for users, which means that bashrc/profile files are not sourced. You can launch a shell for each user with:

```python
c.LocalProcessSpawner.shell_cmd = ['bash', '-l', '-c']
```

which should set up a full login shell as part of launching the server. Do note that this should only be done for semi-trusted users where users being able to take full control over their servers is acceptable (i.e. that you trust users not to mess with each other by sending them malicious links, etc.).

**Comment by chack05** [comment]
@minrk Thank you very much. Why does it work out of the box with the Python 3 kernel, btw?

**Comment by minrk** [comment]
I wouldn't expect it to. Maybe you have something extra in your python3 kernelspec? What's `jupyter kernelspec list` and check the contents of `kernel.json` in each of those directories to see if there's anything suspicious.

**Comment by chack05** [comment]
@minrk Unfortunately, setting `c.LocalProcessSpawner.shell_cmd = ['bash', '-l', '-c']` didn't work.

The contents of the kernel.json files actually aren't that much different:

```
{
 "argv": [
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
}
```
```
{
 "display_name": "Python 2",
 "language": "python",
 "argv": [
  "/usr/bin/python2",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ]
}
```
The only difference seems to be the binary path.

**Comment by chack05** [comment]
@minrk About Python 2 vs Python 3: It seems I was wrong, neither load the environment variables.
It just seemed that way because Python 3 kernels apparently take seriously longer to set up, so it seemed there was no error message for imports of packages from the path.

**Comment by chack05** [comment]
I mitigated the issue by creating a _system wide_ IPython configuration in `/etc/ipython/ipython_config.py`.

For Python 3 kernel users, the file name seems to be `ipython_notebook_config.py` instead. Global ipython configs are apparently missing in the documentation and are only loaded, if the user hasn't create a local configuration file. I got that information from another issue: https://github.com/jupyterhub/jupyterhub/issues/153

There I added PATH and PYTHONPATH similar as to described in [here](https://www.lucypark.kr/blog/2013/02/10/when-python-imports-and-ipython-does-not/) by inserting the following setting into the configuration file:

```
c.InteractiveShellApp.exec_lines = [
  'import sys; sys.path.append("/your/pythonpath/here");import os;os.environ["PATH"]+= os.pathsep + "/your/path/here"'
]
```

Best regards.
